### ISSUE TITLE
Really, really bizarre issue with xontribs that use signal

### ISSUE THREAD

**Issue by kirbyfan64** [xonsh/xonsh#1481]
This is really weird.

To reproduce:
- Create a new `xontrib` directory.
- Create a file `xontrib/bug.py` containing a single line: `import xonsh.__main__`.
- Create `setup.py` containing:
  
  ```
  from distutils.core import setup
  
  setup(
  name='xontrib-bug',
  packages=['xontrib'],
  package_dir={'xontrib': 'xontrib'},
  )
  ```
- Run `python setup.py install`.

Now, if you run `xontrib load bug` twice, _this_ happens:

``` python
ryan@DevPC-LX /media/ryan/stuff/xonsh/xontrib-bug master $ xontrib load bug
ValueError: signal only works in main thread
ryan@DevPC-LX ~ master $ pwd # Why does the prompt say just `~`?
/media/ryan/stuff/xonsh/xontrib-bug
ryan@DevPC-LX ~ master $ xontrib load bug # This will crash Xonsh.
xonsh: For full traceback set: $XONSH_SHOW_TRACEBACK = True
AttributeError: 'module' object has no attribute '__xonsh_config__'
xonsh: For full traceback set: $XONSH_SHOW_TRACEBACK = True
AttributeError: 'module' object has no attribute '__xonsh_history__'
Traceback (most recent call last):
  File "/media/ryan/stuff/anaconda/bin/xonsh", line 9, in <module>
    load_entry_point('xonsh==0.3.0', 'console_scripts', 'xonsh')()
  File "/home/ryan/stuff/anaconda/lib/python3.4/site-packages/xonsh/main.py", line 254, in main
    shell.cmdloop()
  File "/home/ryan/stuff/anaconda/lib/python3.4/site-packages/xonsh/ptk/shell.py", line 120, in cmdloop
    self.default(line)
  File "/home/ryan/stuff/anaconda/lib/python3.4/site-packages/xonsh/base_shell.py", line 168, in default
    self._append_history(inp=src, ts=[ts0, ts1], tee_out=tee.getvalue())
  File "/home/ryan/stuff/anaconda/lib/python3.4/site-packages/xonsh/base_shell.py", line 258, in _append_history
    hist = builtins.__xonsh_history__  # pylint: disable=no-member
AttributeError: 'module' object has no attribute '__xonsh_history__'
Error in atexit._run_exitfuncs:
Traceback (most recent call last):
  File "/home/ryan/stuff/anaconda/lib/python3.4/site-packages/xonsh/built_ins.py", line 744, in _lastflush
    builtins.__xonsh_history__.flush(at_exit=True)
AttributeError: 'module' object has no attribute '__xonsh_history__'
Error in atexit._run_exitfuncs:
Traceback (most recent call last):
  File "/home/ryan/stuff/anaconda/lib/python3.4/site-packages/xonsh/built_ins.py", line 744, in _lastflush
    builtins.__xonsh_history__.flush(at_exit=True)
AttributeError: 'module' object has no attribute '__xonsh_history__'
```

**Comment by scopatz** [comment]
This is because the xontrib command is itself run on a background thread, allowing things like backgrounding etc.  This is coupled with the fact that signal must be used on the main thread.  We could fix this by making the xontrib command happen on the main thread with the `@foreground` decorator.  

Seems like we'd miss the oppurtunity to do real job control with the xontrib command, but I am not sure if people are using that...  Thoughts?

**Comment by kirbyfan64** [comment]
Well, I'm not necessarily saying this should work. I'm just saying that the effects of it not working should be less bizarre (e.g. the random removal of the current directory from the prompt, the second run causing Xonsh to crash).

**Comment by scopatz** [comment]
I think it would be more work to figure out the bizarreness than to fix it :)

**Comment by scopatz** [comment]
Potential fix in #2257

**Comment by melund** [comment]
@kirbyfan64. #2257 is merged in. Please reopen if it doesn't solve your problem.
